### ISSUE TITLE
Possible Bug in package scoped fixture teardown

### ISSUE THREAD

**Issue by s0undt3ch** [pytest-dev/pytest#4684]
Please consider the following diff against PyTest 4.1:
```diff
diff --git a/testing/python/fixture.py b/testing/python/fixture.py
index b6692ac9..7b59cef0 100644
--- a/testing/python/fixture.py
+++ b/testing/python/fixture.py
@@ -3792,7 +3792,11 @@ class TestScopeOrdering(object):
             ├── sub1
             │   ├── __init__.py
             │   ├── conftest.py
-            │   └── test_1.py
+            │   ├── test_1.py
+            │   └── sub1_2
+            │       ├── __init__.py
+            │       ├── conftest.py
+            │       └── test_1_2.py
             └── sub2
                 ├── __init__.py
                 ├── conftest.py
@@ -3824,6 +3828,30 @@ class TestScopeOrdering(object):
         """
             )
         )
+        sub1_2 = sub1.mkdir("sub1_2")
+        sub1_2.ensure("__init__.py")
+        sub1_2.join("conftest.py").write(
+            textwrap.dedent(
+                """\
+            import pytest
+            from ... import values
+            @pytest.fixture(scope="package")
+            def fix():
+                values.append("pre-sub1_2")
+                yield values
+                assert values.pop() == "pre-sub1_2"
+        """
+            )
+        )
+        sub1_2.join("test_1_2.py").write(
+            textwrap.dedent(
+                """\
+            from ... import values
+            def test_1_2(fix):
+                assert values == ["pre-sub1_2"]
+        """
+            )
+        )
         sub2 = root.mkdir("sub2")
         sub2.ensure("__init__.py")
         sub2.join("conftest.py").write(
@@ -3849,7 +3877,103 @@ class TestScopeOrdering(object):
             )
         )
         reprec = testdir.inline_run()
-        reprec.assertoutcome(passed=2)
+        reprec.assertoutcome(passed=3)
+
+    def test_multiple_packages_inherit(self, testdir):
+        """Complex test involving multiple package fixtures. Make sure teardowns
+        are executed in order.
+        .
+        └── root
+            ├── __init__.py
+            ├── sub1
+            │   ├── __init__.py
+            │   ├── conftest.py
+            │   ├── test_1.py
+            │   └── sub1_2
+            │       ├── __init__.py
+            │       ├── conftest.py
+            │       └── test_1_2.py
+            └── sub2
+                ├── __init__.py
+                ├── conftest.py
+                └── test_2.py
+        """
+        root = testdir.mkdir("root")
+        root.join("__init__.py").write("values = []")
+        sub1 = root.mkdir("sub1")
+        sub1.ensure("__init__.py")
+        sub1.join("conftest.py").write(
+            textwrap.dedent(
+                """\
+            import pytest
+            from .. import values
+            @pytest.fixture(scope="package")
+            def fix():
+                values.append("pre-sub1")
+                yield values
+                assert values.pop() == "pre-sub1"
+        """
+            )
+        )
+        sub1.join("test_1.py").write(
+            textwrap.dedent(
+                """\
+            from .. import values
+            def test_1(fix):
+                assert values == ["pre-sub1"]
+        """
+            )
+        )
+        sub1_2 = sub1.mkdir("sub1_2")
+        sub1_2.ensure("__init__.py")
+        sub1_2.join("conftest.py").write(
+            textwrap.dedent(
+                """\
+            import pytest
+            from ... import values
+            @pytest.fixture(scope="package")
+            def fix(fix):
+                values.append("pre-sub1_2")
+                yield values
+                assert values.pop() == "pre-sub1_2"
+        """
+            )
+        )
+        sub1_2.join("test_1_2.py").write(
+            textwrap.dedent(
+                """\
+            from ... import values
+            def test_1_2(fix):
+                assert values == ["pre-sub1", "pre-sub1_2"]
+        """
+            )
+        )
+        sub2 = root.mkdir("sub2")
+        sub2.ensure("__init__.py")
+        sub2.join("conftest.py").write(
+            textwrap.dedent(
+                """\
+            import pytest
+            from .. import values
+            @pytest.fixture(scope="package")
+            def fix():
+                values.append("pre-sub2")
+                yield values
+                assert values.pop() == "pre-sub2"
+        """
+            )
+        )
+        sub2.join("test_2.py").write(
+            textwrap.dedent(
+                """\
+            from .. import values
+            def test_2(fix):
+                assert values == ["pre-sub2"]
+        """
+            )
+        )
+        reprec = testdir.inline_run()
+        reprec.assertoutcome(passed=3)
 
 
 def test_call_fixture_function_error():
```
The difference in the tests is that on `test_multiple_packages`, there's no fixture reuse/extend, while on `test_multiple_packages_inherit` I "extend" the `fix` fixture.

On `test_multiple_packages_inherit`, at teardown, I expect sub_1_2 to cleanup first, then sub_1, but the test fails at sub_2 because it still contains the item added in sub_1(which didn't ran it's teardown code yet?

**Comment by s0undt3ch** [comment]
Unless, of course, my ordering assumptions are wrong.

**Comment by Zac-HD** [comment]
I don't know enough about fixtures to comment on the details, but you'd probably have better luck opening a pull request to propose changes.

**Comment by s0undt3ch** [comment]
I shouldn't propose changes if the behaviour is what's expected, hence this pre-PR issue.
Closed and unanswered...

**Comment by s0undt3ch** [comment]
Also, the above is just proposing a test shiv will fail, the fix is yet unknown. Don't even know if its desirable, if it is, this shouldn't have been closed.

**Comment by nicoddemus** [comment]
@s0undt3ch indeed looks like a bug, thanks for reporting it.

**Comment by RomanPylypenkoDevPro** [comment]
I have the similar issue with "scope=package" when I want to group my test in a few sub-folders using one common conftest.
In the case below, the teardown of package_1/conftest runs at the end of the session just after teardown of package_2/conftest:
```
        .
        └── root
            ├── __init__.py
            ├── package_1
            │   ├── __init__.py
            │   ├── conftest.py
            │   └── sub_package1
            │       ├── __init__.py
            │       └── test_1.py
            └── package_2
                 ├── __init__.py
                 ├── conftest.py
                 └── test_2.py
```
But at the same time the issue is not reproduced if there are no any nested tests in sub-packages like:
```
        .
        └── root
            ├── __init__.py
            ├── package_1
            │   ├── __init__.py
            │   ├── conftest.py
            │   └── test_1.py
            └── package_2
                 ├── __init__.py
                 ├── conftest.py
                 └── test_2.py
```
Now the order of fixtures execution is correct.
Full code of example with issue reproduced: https://github.com/RomanPylypenkoDevPro/Sandbox

I hope this info will be useful.

**Comment by s0undt3ch** [comment]
I've been trying to address this issue.
My current idea is that each `FixtureDef` will store each nodeid where it's used.
Then instead of relying on the schedule finalization routine, we check if there aren't any more node IDs for said `FixtureDef` and in said case, we run the fixture finalize method immediately.

Which at least suggests how fixture teardown should be, ie, if not going to be used anymore, finalize it.

Now, it breaks a lot of tests, even an acceptance test which just passes a tempdir to `pytest.main()` and is supposed to rerun an exit code indicating no tests were collected...

Would you agree that this fixture termination is the desired one but we've just been relying on deferred termination routines?

Is this the wrong approach?

**Comment by shughes-uk** [comment]
Hi there I ran into this issue today. Thanks @RomanPylypenkoDevPro for explaining it. I guess the workaround for now is to avoid sub packages?

**Comment by s0undt3ch** [comment]
This is a hack but [this](https://github.com/saltstack/salt/blob/master/tests/conftest.py#L308) and [this](https://github.com/saltstack/salt/blob/master/tests/conftest.py#L377) is what we're doing to force fixture termination once no test is using it.

**Comment by s0undt3ch** [comment]
Is it fair to say that if we are in package scope A and we enter package B which is a sub-package of A, any fixtures in the package A scope should **not** be teardown?
I want to work on this but guidelines on where PyTest would like this to go is preferable in order not to waste time.

I still think it's useful to teardown fixtures when they are no longer going to be used, but this "usage tracking" might not be preferable.

**Comment by s0undt3ch** [comment]
New diff for tests against current master https://gist.github.com/s0undt3ch/021d86c62ab2c4824abea23b1cc9c5f5#file-fixtures-diff-L16

`test_multiple_packages_complex_2` fails of course

**Comment by thenx** [comment]
Guess the bug is still there.

Env:
```
─➤  pip list 
Package             Version                   Editable project location
------------------- ------------------------- ----------------------------
attrs               22.2.0
exceptiongroup      1.1.0
iniconfig           2.0.0
packaging           23.0
pip                 22.3.1
pluggy              1.0.0
pytest              7.2.1
pytest-asyncio      0.20.3
setuptools          67.0.0
tomli               2.0.1
wheel               0.38.4

```



Folder structure:
```
─➤  tree tests                
tests
├── __init__.py
├── example
│   ├── __init__.py
│   ├── conftest.py
│   └── inner
│       ├── __init__.py
│       └── test_inner.py
└── outer
    ├── __init__.py
    └── test_outer.py

```

Content of `conftest.py`:
```python
import asyncio
from collections.abc import AsyncIterator, Iterable

import pytest


@pytest.fixture(scope="package")
def event_loop() -> Iterable[asyncio.AbstractEventLoop]:
    policy = asyncio.get_event_loop_policy()
    loop = policy.new_event_loop()
    yield loop
    loop.close()


@pytest.fixture(scope="package")
async def async_package_fixture() -> AsyncIterator[None]:
    yield

```

Content of `test_inner.py`:
```python
async def test_inner(async_package_fixture: None) -> None:
    pass

```

Content of `test_outer.py`:
```python
async def test_outer() -> None:
    pass

```

Output of `pytest --setup-show`:
```
─➤  pytest --setup-show
=============================================================================== test session starts ===============================================================================
platform darwin -- Python 3.10.9, pytest-7.2.1, pluggy-1.0.0
rootdir: <rootdir>, configfile: pyproject.toml, testpaths: tests
plugins: asyncio-0.20.3
asyncio: mode=auto
collected 2 items                                                                                                                                                                 

tests/example/inner/test_inner.py 
  SETUP    P event_loop
  SETUP    P async_package_fixture (fixtures used: event_loop)
        tests/example/inner/test_inner.py::test_inner (fixtures used: async_package_fixture, event_loop, request).
tests/outer/test_outer.py 
        SETUP    F event_loop
        tests/outer/test_outer.py::test_outer (fixtures used: event_loop).
        TEARDOWN F event_loop
  TEARDOWN P async_package_fixture
  TEARDOWN P event_loopE

===================================================================================== ERRORS ======================================================================================
_________________________________________________________________________ ERROR at teardown of test_outer _________________________________________________________________________

    def finalizer() -> None:
        """Yield again, to finalize."""
    
        async def async_finalizer() -> None:
            try:
                await gen_obj.__anext__()
            except StopAsyncIteration:
                pass
            else:
                msg = "Async generator fixture didn't stop."
                msg += "Yield only once."
                raise ValueError(msg)
    
>       event_loop.run_until_complete(async_finalizer())

.venv/lib/python3.10/site-packages/pytest_asyncio/plugin.py:297: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../.pyenv/versions/3.10.9/lib/python3.10/asyncio/base_events.py:624: in run_until_complete
    self._check_closed()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_UnixSelectorEventLoop running=False closed=True debug=False>

    def _check_closed(self):
        if self._closed:
>           raise RuntimeError('Event loop is closed')
E           RuntimeError: Event loop is closed

../../.pyenv/versions/3.10.9/lib/python3.10/asyncio/base_events.py:515: RuntimeError
============================================================================= short test summary info =============================================================================
ERROR tests/outer/test_outer.py::test_outer - RuntimeError: Event loop is closed
=========================================================================== 2 passed, 1 error in 0.06s ============================================================================
sys:1: RuntimeWarning: coroutine '_wrap_asyncgen_fixture.<locals>._asyncgen_fixture_wrapper.<locals>.finalizer.<locals>.async_finalizer' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback

```

Guess the main info is the following:
```
tests/example/inner/test_inner.py 
  SETUP    P event_loop
  SETUP    P async_package_fixture (fixtures used: event_loop)
        tests/example/inner/test_inner.py::test_inner (fixtures used: async_package_fixture, event_loop, request).
tests/outer/test_outer.py 
        SETUP    F event_loop
        tests/outer/test_outer.py::test_outer (fixtures used: event_loop).
        TEARDOWN F event_loop
  TEARDOWN P async_package_fixture
  TEARDOWN P event_loopE
```
There's no teardown for package-scoped event_loop before setup/teardown of function-scoped event_loop. The consequence of it is an attempt to close closed event_loop.

**Comment by RonnyPfannschmidt** [comment]
Currently we have the issue that multi level package scope is practically undefined and not safely fixable without major reworks

**Comment by bluetech** [comment]
I haven't tested the exact scenarios, but I believe this should be fixed with #11646. Please test `pip install pytest==8.0.0rc1` and let us know if not!

**Comment by thenx** [comment]
Confirm that my example is resolved in 8.0.0rc1

**Comment by Dianne087** [comment]
import pytest
import os
import shutil

@pytest.fixture(scope="session")
def cleanup_environment():
    temp_paths = []  # List to track temporary paths

    yield temp_paths  # Provide the list to tests

    # Sort paths by directory depth (deepest first)
    temp_paths_sorted = sorted(temp_paths, key=lambda p: p.count(os.sep), reverse=True)

    for temp_path in temp_paths_sorted:
        if os.path.exists(temp_path):
            try:
                if os.path.isfile(temp_path):
                    os.remove(temp_path)
                elif os.path.isdir(temp_path):
                    shutil.rmtree(temp_path, ignore_errors=True)
            except Exception as e:
                print(f"Failed to remove {temp_path}: {e}")

    for temp_path in temp_paths_sorted:
        parent_dir = os.path.dirname(temp_path)
        while parent_dir and os.path.exists(parent_dir) and not os.listdir(parent_dir):
            try:
                os.rmdir(parent_dir)
                parent_dir = os.path.dirname(parent_dir)
            except Exception as e:
                print(f"Failed to remove parent directory {parent_dir}: {e}")
                break
Here are the test cases:

Works (Directories and files are cleaned up):

def test_create_and_delete_file(cleanup_environment, tmpdir):
    temp_file = tmpdir.join("temp_file.txt")
    temp_file.write("Temporary content")
    cleanup_environment.append(str(temp_file))
    assert temp_file.check(file=True)

def test_create_and_delete_dir(cleanup_environment, tmpdir):
    temp_dir = tmpdir.mkdir("temp_dir")
    cleanup_environment.append(str(temp_dir))
    assert temp_dir.check(dir=True)
Fails (Directories from these tests are not fully cleaned up):

@pytest.mark.parametrize("test_case", ["case1", "case2"])
def test_parameterized_cases(cleanup_environment, tmpdir, test_case):
    temp_dir = tmpdir.mkdir(f"temp_dir_{test_case}")
    cleanup_environment.append(str(temp_dir))
    assert temp_dir.check(dir=True)
In the parameterized test cases (case1 and case2), the directories created are not fully deleted, even though they are appended to cleanup_environment.

How can I ensure that all directories, including those from parameterized tests, are cleaned up properly?

**Comment by Dianne087** [comment]
<advance>optimeser on remoteimport pytest
import os
import shutil

@pytest.fixture(scope="session")
def cleanup_environment():
    temp_paths = []  # List to track temporary paths

    yield temp_paths  # Provide the list to tests

    # Sort paths by directory depth (deepest first)
    temp_paths_sorted = sorted(temp_paths, key=lambda p: p.count(os.sep), reverse=True)

    for temp_path in temp_paths_sorted:
        if os.path.exists(temp_path):
            try:
                if os.path.isfile(temp_path):
                    os.remove(temp_path)
                elif os.path.isdir(temp_path):
                    shutil.rmtree(temp_path, ignore_errors=True)
            except Exception as e:
                print(f"Failed to remove {temp_path}: {e}")

    for temp_path in temp_paths_sorted:
        parent_dir = os.path.dirname(temp_path)
        while parent_dir and os.path.exists(parent_dir) and not os.listdir(parent_dir):
            try:
                os.rmdir(parent_dir)
                parent_dir = os.path.dirname(parent_dir)
            except Exception as e:
                print(f"Failed to remove parent directory {parent_dir}: {e}")
                break
Here are the test cases:

Works (Directories and files are cleaned up):

def test_create_and_delete_file(cleanup_environment, tmpdir):
    temp_file = tmpdir.join("temp_file.txt")
    temp_file.write("Temporary content")
    cleanup_environment.append(str(temp_file))
    assert temp_file.check(file=True)

def test_create_and_delete_dir(cleanup_environment, tmpdir):
    temp_dir = tmpdir.mkdir("temp_dir")
    cleanup_environment.append(str(temp_dir))
    assert temp_dir.check(dir=True)
Fails (Directories from these tests are not fully cleaned up):

@pytest.mark.parametrize("test_case", ["case1", "case2"])
def test_parameterized_cases(cleanup_environment, tmpdir, test_case):
    temp_dir = tmpdir.mkdir(f"temp_dir_{test_case}")
    cleanup_environment.append(str(temp_dir))
    assert temp_dir.check(dir=True)
In the parameterized test cases (case1 and case2), the directories created are not fully deleted, even though they are appended to cleanup_environment.

How can I ensure that all directories, including those from parameterized tests, are cleaned up properly?
